### PR TITLE
Reflection caching

### PR DESCRIPTION
**Pull request by dvoet**
The crux of this change is in class CachedType
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [ ] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [ ]  sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Squash commits, rebase if necessary
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Merge to develop 
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: Check configuration files in Jenkins in case they need changes
- [ ] **Submitter**: Verify swagger UI on dev server still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 1]
```diff
@@ -77,7 +77,107 @@ object EdgeSchema {
...
 
+/**
+ * This class is a wrapper around scala.reflect.api.Type that caches results of calls to the Type so
+ * we don't need to go through reflection for future calls. Caching is done through lazy vals in most
+ * cases but TrieMaps are used in places where results build over time.
+ * @param tpe
+ * @tparam T
+ */
+class CachedType[T](val tpe: Type) {
+  def this()(implicit ttag: TypeTag[T]) = this(typeOf[T])
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: What does this mean? I don't understand. Further down you're declaring types as `CachedType[_]`, and I don't understand that either.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 2]
```diff
@@ -77,7 +77,107 @@ object EdgeSchema {
...
 }
 
+/**
+ * This class is a wrapper around scala.reflect.api.Type that caches results of calls to the Type so
+ * we don't need to go through reflection for future calls. Caching is done through lazy vals in most
+ * cases but TrieMaps are used in places where results build over time.
+ * @param tpe
+ * @tparam T
+ */
+class CachedType[T](val tpe: Type) {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: I'd prefer a separate file for the caching.  `GraphDAO` is pretty big.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 3]
```diff
@@ -348,31 +440,32 @@ trait GraphDAO {
...
-      saveProperty(typeOf[DateTime], "lastModified", DateTime.now, wsc.get.workspaceVertex, wsc, graph)
+      saveProperty(cachedTypeOf[DateTime], "lastModified", DateTime.now, wsc.get.workspaceVertex, wsc, graph)
     }
   }
 
   // a function to uniquely identify a graph object relative to siblings, by checking the idFields for equality
-  private def domainObjectFilterFn(tpe: Type, obj: DomainObject)(vertex: Vertex): Boolean = {
-    obj.idFields.forall { field => obj.getFieldValue(tpe, field) == vertex.getProperty(field) }
+  private def domainObjectFilterFn(tpe: CachedType[_], obj: DomainObject): Vertex => Boolean = {
+    val idFieldValues = obj.idFields.map(field => (field, obj.getFieldValue(tpe.tpe, field)))
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: why this change?

Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: this is a performance optimization. In the old way `obj.getFieldValue` is called on every comparison. This uses reflection and is costly. The new way gets the values once.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala (hunk 1 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: we talked about this, but for posterity:
this line declares a no-arg constructor and gets the type from an implicit TypeTag parameter. This class has a type parameter so that the compiler can differentiate between the different CachedType instances. The rest of the code does not care about the type thus the [_] everywhere.

File: src/main/config/rawls.conf.ctmpl [hunk 4]
```diff
@@ -1,9 +1,8 @@
 {{with $environment := env "ENVIRONMENT"}}{{$keyname := printf "secret/dsde/%s/rawls/rawls.conf" $environment}}{{with vault $keyname}}
 orientdb {
-  server = "{{.Data.orientdb_server}}"
+  connectionUrl = "{{.Data.orientdb_connection_url}}"
```
Comment (helgridly) [inline src/main/config/rawls.conf.ctmpl:3]: seems to me this db stuff wants to be in a separate commit

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala (hunk 1 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: ^ can you add this as a comment to the code

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala (hunk 3 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: still not following this change. the results aren't stored outside the scope of this function so how does this help?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 5]
```diff
@@ -348,31 +440,32 @@ trait GraphDAO {
...
   }
 
-  def saveSubObject[T <: DomainObject :TypeTag :ClassTag](propName: String, obj: T, vertex: Vertex, wsc: Option[WorkspaceContext], graph: Graph): Vertex = {
-    saveSubObject(typeOf[T], propName, obj, vertex, wsc, graph)
+  def saveSubObject[T <: DomainObject :TypeTag :ClassTag :CachedType](propName: String, obj: T, vertex: Vertex, wsc: Option[WorkspaceContext], graph: Graph): Vertex = {
+    saveSubObject(cachedTypeOf[T], propName, obj, vertex, wsc, graph)
   }
 
-  def saveSubObject(tpe: Type, propName: String, obj: DomainObject, vertex: Vertex, wsc: Option[WorkspaceContext], graph: Graph): Vertex = {
+  def saveSubObject(tpe: CachedType[_], propName: String, obj: DomainObject, vertex: Vertex, wsc: Option[WorkspaceContext], graph: Graph): Vertex = {
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: can we make a typedef for `CachedType[_]` so i don't have to see that offensive `_` all over the code?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphEntityDAO.scala [hunk 6]
```diff
@@ -11,6 +11,7 @@ import scala.collection.JavaConversions._
 import scala.collection.JavaConverters._
 
 import org.broadinstitute.dsde.rawls.model.AttributeConversions._
+import CachedTypes._
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphEntityDAO.scala:14]: what happens if we forget to do this?
- does it compile anyway and fail mysteriously? (bad)
- is it obvious that this is the way to fix it?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala [hunk 7]
```diff
@@ -178,6 +178,8 @@ class HttpGoogleServicesDAO(
 
         bucketName
       }
+    } recover {
+      case t: HttpResponseException if t.getStatusCode == 409 => bucketName
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:183]: this doesn't look like it belongs in this PR

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 8]
```diff
@@ -236,12 +228,10 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: DataSource,
...
         }
-      }
-    })
 
-    // collect the result
-    eventualResponses map (r => RequestComplete(StatusCodes.OK, r.collect { case Some(x) => x }))
-  }
+        RequestComplete(StatusCodes.OK, responses.collect { case Some(x) => x })
+      }
+    }
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:234]: nor any of this

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala (hunk 2 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: +1

**Comment by helgridly** [comment]
summary of feedback from inline comments:
- stuff to break out into another PR
- pull out caching stuff into another file
- more commentary on CachedType
- typedef for CachedType[_]
- more questions on domainObjectFilterFn
- questions about what happens if we forget to import the implicits in other files

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphEntityDAO.scala (hunk 6 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphEntityDAO.scala:14]: if it is forgotten it will not compile and complain that it can't find an implicit

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala (hunk 3 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: After reading through this a few times, I believe this decreases the number of `getFieldValue` calls from once per vertex in the traversal to only once per traversal.

Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: the old was was a curried function which waited for both sets of parameters to be defined before evaluation. The new way is a function that returns a function. The first function evaluates the fields which are embedded in the returned function. In effect, the values of the fields are stored outside the scope of the returned function.

Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: how? `obj.getFieldValue` isn't cached anywhere at all

Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: ah, i see. some commentary here might be good in order to clear this up, since it's confused two people
"this returns a filter function where the id field values have already been looked up, rather than calling `getFieldValue` every time the filter function is run"

**Comment by helgridly** [comment]
Very naughty. There was no :+1: .

**Comment by dvoet** [comment]
sorry, I addressed all the things except pulling out the separate PRs... thought that was good enough

**Comment by helgridly** [comment]
this is why we have the tickboxes :(
